roles/edpm_nova/scripts/derive_pci_passthrough_whitelist.py

```diff
@@ -0,0 +1,587 @@
+) -> None:
+    if prop == ANY:
+        return REGEX_ANY
+    try:
+        v = int(prop, 16)
+    except ValueError:
+        raise InvalidConfigException('Invalid PCI address specififed {!r}'.format(prop))
+    if v > maxval:
+        raise InvalidConfigException('PCI address specififed {!r} is out of range'.format(prop))
+    return hex_value % v
+
+
+def get_pciaddr_dict_from_usraddr(pci_addr: str):
+    """Convert PCI address in STRING format to DICT
+    (this is done for uniformity in PCI address)
+    """
+    pci_dict = {}
+    dbs, sep, func = pci_addr.partition('.')
+    pci_dict['function'] = ANY
+    if func:
+        func = func.strip()
+        pci_dict['function'] = func
+    if dbs:
+        dbs_fields = dbs.split(':')
+        if len(dbs_fields) > 3:
+            raise InvalidConfigException('Invalid PCI address specififed {!r}'.format(pci_addr))
+        # If we got a partial address like ":00.", we need to turn this
+        # into a domain of ANY, a bus of ANY, and a slot of 00. This code
+        # allows the address,bus and/or domain to be left off
+        dbs_all = [ANY] * (3 - len(dbs_fields))
+        dbs_all.extend(dbs_fields)
+        dbs_checked = [s.strip() or ANY for s in dbs_all]
+
+        ''' domain, bus, slot = dbs_checked '''
+        pci_dict['domain'], pci_dict['bus'], pci_dict['slot'] = dbs_checked
+
+    pci_dict['domain'] = get_pci_field_val(pci_dict['domain'], MAX_DOMAIN, '%04x')
+    pci_dict['slot'] = get_pci_field_val(pci_dict['slot'], MAX_SLOT, '%02x')
+    pci_dict['bus'] = get_pci_field_val(pci_dict['bus'], MAX_BUS, '%02x')
+    pci_dict['function'] = get_pci_field_val(pci_dict['function'], MAX_FUNC, '%1x')
+    return pci_dict
+
+
+def get_pci_regex_pattern(config_regex: str, size: int, maxval: int, hex_value: str):
+    if config_regex in [ANY, REGEX_ANY]:
+        config_regex = "[0-9a-fA-F]{%d}" % size
+
+    try:
+        re.compile(config_regex)
+    except re.error:
+        msg = "Invalid regex pattern identified %s" % config_regex
+        raise InvalidConfigException(msg)
+    try:
+        v = int(config_regex, 16)
+    except ValueError:
+        return config_regex
+    if v > maxval:
+        msg = "Invalid pci address"
+        raise InvalidConfigException(msg)
+    return hex_value % v
+
+
+def get_user_regex_from_addrdict(addr_dict):
+    if isinstance(addr_dict, dict):
+        domain_regex = get_pci_regex_pattern(addr_dict['domain'], 4, MAX_DOMAIN, '%04x')
+        bus_regex = get_pci_regex_pattern(addr_dict['bus'], 2, MAX_BUS, '%02x')
+        slot_regex = get_pci_regex_pattern(addr_dict['slot'], 2, MAX_SLOT, '%02x')
+        function_regex = get_pci_regex_pattern(addr_dict['function'], 1, MAX_FUNC, '%1x')
+
+        user_address_regex = '%s:%s:%s.%s' % (domain_regex,
+                                              bus_regex, slot_regex,
+                                              function_regex)
+        return user_address_regex
+    else:
+        return None
```

Comment:
Function can return `None` or a string. Consuming functions need a check, or a rewrite.

Reply:
It is handled: https://github.com/openstack-archive/tripleo-heat-templates/blob/stable/wallaby/deployment/neutron/derive_pci_passthrough_whitelist.py#L370